vyper/ast/nodes.py

```diff
@@ -15,34 +14,107 @@
 )
 
 BASE_NODE_ATTRIBUTES = (
-    'node_id',
-    'source_code',
+    'ast_type',
     'col_offset',
-    'lineno',
     'end_col_offset',
     'end_lineno',
-    'src'
+    'lineno',
+    'node_id',
+    'source_code',
+    'src',
 )
+DICT_AST_SKIPLIST = ('source_code', )
+
+
+def get_node(ast_struct: typing.Union[typing.Dict, python_ast.AST]) -> "VyperNode":
+    """
+    Converts an AST structure to a Vyper AST node.
+
+    This is a recursive call, all child nodes of the input value are also
+    converted to Vyper nodes.
+
+    Attributes
+    ----------
+    ast_struct: (dict, AST)
+        Annotated python AST node or Vyper AST dict to generate the node from.
+
+    Returns
+    -------
+    VyperNode
+        The generated AST object.
+    """
+    if not isinstance(ast_struct, dict):
+        ast_struct = ast_struct.__dict__
+
+    vy_class = getattr(sys.modules[__name__], ast_struct['ast_type'], None)
+
+    if vy_class is None:
+        raise SyntaxException(
+            f"Invalid syntax (unsupported '{ast_struct['ast_type']}'' Python AST node).",
+            ast_struct
+        )
+
+    return vy_class(**ast_struct)
+
+
+def _to_node(value):
+    if isinstance(value, (dict, python_ast.AST)):
+        return get_node(value)
+    return value
+
+
+def _to_dict(value):
+    if isinstance(value, VyperNode):
+        return value.to_dict()
+    return value
 
 
 class VyperNode:
+    """
+    Base class for all vyper AST nodes.
+
+    Vyper nodes are generated from, and closely resemble, their Python counterparts.
+
+    Attributes
+    ----------
+    __slots__ : Tuple
+        Allowed field names for the node.
+    _only_empty_fields : Tuple
+        Field names that, if present, must be set to None or a SyntaxException is
+        raised. This attribute is used to exclude syntax that is valid in Python
+        but not in vyper.
```

Comment:
Why are these necessary? Because (for now) we're using Python AST nodes as a base class?

Reply:
Yeah this is to block things like `else` within a for loop, or any other syntax that's valid in python but not in vyper. Once we do our own parsing it can go.